### PR TITLE
Worked toward full user maps support.

### PR DESCRIPTION
**Pull request by Bertram25**
Enjoy! :)

### CODE REVIEW

**Comment by Bertram25** [comment]
@hwoarangmy Your turn (for once) ;)

**Comment by Bertram25** [comment]
Rebased.

**Comment by hwoarangmy** [comment]
@Bertram25 ok, I will review that ASAP. Nice to see you back in the coding arena ^^

**Comment by Bertram25** [comment]
> @Bertram25 ok, I will review that ASAP. Nice to see you back in the coding arena ^^

Thanks. :) If this PR gets accepted, I'll be able to actually start working on the New Map Menu + Map properties dialog. :)

**Comment by hwoarangmy** [comment]
> Thanks. :) If this PR gets accepted, I'll be able to actually start working on the New Map Menu + Map properties dialog. :)

Well, feel free to add new commits in this PR (especially since it is about the menus as some other commits in this PR). I don't have any code breaking PR right now ^^

**Comment by Bertram25** [comment]
> Well, feel free to add new commits in this PR (especially since it is about the menus as some other commits in this PR). I don't have any code breaking PR right now ^^

Eh eh. (I'm lucky). The thing is, I overhauled the Main menu buttons, so I'd rather wait for your feed-back about it before keeping at it. Just in case you dislike it for any good reason.

**Comment by hwoarangmy** [comment]
I've tried and I like the idea of adding icons. However, using a window is not very nice IMHO.
I prefer what @knight666 did. You can have a look:
https://github.com/knight666/OpenDungeons/tree/ui-redesign

The general idea is to only display the buttons (as you did) and to display more buttons when one is clicked.

**Comment by hwoarangmy** [comment]
nice re-use of the icons BTW ^^

**Comment by Bertram25** [comment]
> The general idea is to only display the buttons (as you did) and to display more buttons when one is clicked.

Ok, I'll test his fork asap. Are my changes doomed (for the last commit), can I still base on it for now?

**Comment by hwoarangmy** [comment]
> Are my changes doomed (for the last commit), can I still base on it for now?

Well, have a look and tell me what you prefer. I think what he did was pretty nice and IMHO, it would be better to go for something similar. But in any case, I will not do anything in the coming days that will get in your way so you can take the time you need.

**Comment by Bertram25** [comment]
@OpenDungeons/developers 

> Well, have a look and tell me what you prefer. I think what he did was pretty nice and IMHO, it would be better to go for something similar. But in any case, I will not do anything in the coming days that will get in your way so you can take the time you need.

Ok, I've just tested @knight666 's changes . Here is what I like:
- Using simple backgroundless window to make buttons appear/disappear is clever.
- I can adapt my commit to go toward this.
- The golden font effect is cool, and I should retake that part.

But here is what I dislike:
- The UI background chosen. Maybe it's just a POC of course, but I personally think the current one is better.
- Should we absolutely place buttons on the left, or is this possible to keep the central placement and make buttons appear on need, next to the main button?
- I think my sub-menu logic is better. (The choice of sub-menu used) ;P

What precisely do you like/dislike in it? I'll adapt once we agree on something.

**Comment by hwoarangmy** [comment]
Note that @knight666 popped on the forum and said he is still working on the UI. It would be nice to not mess too much with that until he is finished as long as he still wants to work on that.

> The UI background chosen. Maybe it's just a POC of course, but I personally think the current one is better.

For my part, I like the UI he used. IMHO, it looks more like what we can find in similar games. However, If you don't like it, it would be nice to tell him what you don't like in the UI in the forum thread and discuss that. 

> Should we absolutely place buttons on the left, or is this possible to keep the central placement and make buttons appear on need, next to the main button?

Well, AFAIR, we agreed that the goal would be to remove the background image and display an Ogre scene with some of our models (like in summoning wars). If we go for that, placing the buttons on the left (or on top) makes sense.

> I think my sub-menu logic is better. (The choice of sub-menu used) ;P

From what I remember, it was pretty similar. However, @knight666 forked before I added the master server buttons so they are not displayed in his menu. But that's also something you can discuss. You both are skilled in menu designing and it would be cool to find some agreement about those points.

**Comment by knight666** [comment]
Yes, your continued GitHub mentions in my inbox have awakened me from my slumber! ;)

> The UI background chosen. Maybe it's just a POC of course, but I personally think the current one is better.

If you're talking about the wood texture for the floating window: I agree. I'm really not happy with how that turned out. My main concern was to avoid the current "black button on a black window" look we've got going on, but I don't feel I've really nailed the "aristrocratic evil" look I'm going for yet.

> Should we absolutely place buttons on the left, or is this possible to keep the central placement and make buttons appear on need, next to the main button?

All the games I've worked on so far have either had their menu items on the left or on the right of the screen. It makes it easier to add animations between menu screens and to show off a bit of the game in the background. It's not the worst thing in the world to put the buttons in the center, but you may run into trouble if you're also trying to show off the game at the same time. For reference, this is how Thief (2013) did it: https://i.ytimg.com/vi/N22woKav9qU/maxresdefault.jpg. This placement allowed them to show a bit of the game world and get you in the mood for the game. I didn't design this menu, but I _did_ make it work with the mouse. ;) (Lots of overtime was involved. 0/10 would not recommend.)

> I think my sub-menu logic is better. (The choice of sub-menu used) ;P

I'm gonna give it a try tomorrow. I love being wrong!

For the record: y'all do what y'all gotta do. My first order of business will be to merge from upstream into my branch, but I don't expect to be done with my changes soon. I don't mind merging conflicting changes, so if you guys decide to radically alter the layout of the menus, that's fine, let me worry about that. :)

**Comment by hwoarangmy** [comment]
@Bertram25 Sorry for not reviewing but had no time yesterday and there are some changes that are not that straight forward. I will try to have a look tonight.
Concerning your menu, I think it is better than the current one even if we can improve. And since @knight666 already said he likes to fight with conflicting commits, I think we can merge after review. Do you agree ?

**Comment by Bertram25** [comment]
Hey guys, :)

First of all, if I may have sounded harsh or something through my: like/dislike comment, please don't see it as an offense. Just in case. ;)

> Note that @knight666 popped on the forum and said he is still working on the UI. It would be nice to not mess too much with that until he is finished as long as he still wants to work on that.

Sure. I actually messed with it because I felt bad adding one more button on the main menu for the New map Menu, but I definitely wouldn't go against the whole team wish and roadmap.

> Well, AFAIR, we agreed that the goal would be to remove the background image and display an Ogre scene with some of our models (like in summoning wars). If we go for that, placing the buttons on the left (or on top) makes sense.

Ah, makes sense, indeed. The button should definitely go left in that case. Can't wait to see that scene. :D 

@knight666 I think we never talked before: Welcome into the team! :)

> If you're talking about the wood texture for the floating window: I agree. I'm really not happy with how that turned out. My main concern was to avoid the current "black button on a black window" look we've got going on, but I don't feel I've really nailed the "aristrocratic evil" look I'm going for yet.

Ah I see what you mean, I must say I would definitely not merge your work because of that (and working on the UI takes time unfortunately.), but I think aiming at the "aristocratic evil" theme may look good. :)

> For the record: y'all do what y'all gotta do. My first order of business will be to merge from upstream into my branch, but I don't expect to be done with my changes soon. I don't mind merging conflicting changes, so if you guys decide to radically alter the layout of the menus, that's fine, let me worry about that. :)

Well, here, sorry but I must say I disagree. As we are a spare-time project where already 90% (if not more) of the work is done by @hwoarangmy , you shouldn't accept to "work for nothing", ever, or have to handle crushing conflicts too often.
To avoid this, we should import your commits as soon as they are good to go, even if it's one commit at a time, with caring about the fact that they don't place the game in an unfinished or ugly looking state.

I've looked at the 'ui_redesign' branch, and I was at first lost with the commits order. Could you:
git pull --rebase this branch based on the main development branch as a first step?
it should be something like: `git pull --rebase git://github.com/OpenDungeons/OpenDungeons.git development` from your ui_redesign branch.
Once the conflicts (if any) are fixed, you should be able to push --force in your branch, something like: `git push -f origin ui_redesign`

With this, we will have a good base to see what commits are actually pending, and if your create a PR based on that, we'll be able to start commenting those commits and see what actual work is to be done to get the whole thing integrated.

Now for the whole UI concept, @knight666 and @hwoarangmy if you agree, I can change my last commit to make use of the same logic than @knight666 , so that we have something integrable right away before I start working on the new map ui. Ok for you?

**Comment by hwoarangmy** [comment]
> Well, here, sorry but I must say I disagree. As we are a spare-time project where already 90% (if not more) of the work is done by @hwoarangmy , you shouldn't accept to "work for nothing", ever, or have to handle crushing conflicts too often.

From what @knight666 said, I had understood that he doesn't want to stop us because he doesn't know when he will have a chance to work on OD again. Not that he wants to mess with conflicts for the fun ^^

@knight666 However, I think @Bertram25 is pretty right when he says that there is already enough work improving the game to not use our free time handling endless conflicting merges. I understand your will to come with a final PR that will do everything you have in mind. However, it would be better to split your UI changes into several PRs so that you don't have too many conflicts  to deal with. For example, your new icon disposal seems pretty nice to me so you could create a PR using the old buttons and focus on the new buttons in a separate one (that would be perfect since you will probably be the only one to deal with this for now so you should not have conflicts).

**Comment by hwoarangmy** [comment]
> Now for the whole UI concept, @knight666 and @hwoarangmy if you agree, I can change my last commit to make use of the same logic than @knight666 , so that we have something integrable right away before I start working on the new map ui. Ok for you?

yes ^^

**Comment by akien-mga** [comment]
Hi guys :) Sorry I was quite busy with holidays and such ;)

I'll test this PR asap. Regarding the above discussions on the UI, I think that since we now have two UI experts (with @knight666 likely having more professional experience, but @Bertram25 isn't half bad either!), you guys should probably try to work together on a mockup in GIMP or similar to decide what you'd like to have as a final design.

Once everyone agrees on the mockup, I guess @knight666 could be in charge of the implementation in CEGUI with @Bertram25 as advisor (since @Bertram25 can also find tons of other tasks to do in the meantime as he knows the code base very well). WDYT?

**Comment by Bertram25** [comment]
Hey guys, :)

Merry Christmas (even if late) to everybody!! :D 

> I think that since we now have two UI experts (with @knight666 likely having more professional experience, but @Bertram25 isn't half bad either!), you guys should probably try to work together on a mockup in GIMP or similar to decide what you'd like to have as a final design.

Eh eh, you foxy sycophant. ;) In any case, you are right, and no, I won't present myself as an expert, just someone who follows his heart using his end-user feelings and point of view, even if I'm not doing my first GUI here.

I also do think a common vision of the main menu and so on is the way to go, and I'll look for the corresponding thread on the forums after the next changes. But let's iterate because I don't believe in the fact we'll find the right vision in one pass. Nobody usually can, after all.

So first, I'll adapt the main menu so that it fits current @knight666's vision of it. The goal is to integrate this. Then, maybe @knight666 will be able to port the functional parts of his work in smaller PRs?

**Comment by Bertram25** [comment]
@hwoarangmy @knight666 This should be ok now.

**Comment by hwoarangmy** [comment]
@Bertram25 I've tried your PR and it worsk well once the said lines are removed. However, could you please use the same buttons as the main menu for the sub menus ? It looks a bit weird IMHO to not use the same.

**Comment by Bertram25** [comment]
> However, could you please use the same buttons as the main menu for the sub menus ? It looks a bit weird IMHO to not use the same.

Ok, give me something like an hour or so. :)

**Comment by hwoarangmy** [comment]
> Ok, give me something like an hour or so. :)

No pb, you don't have to rush ^^

**Comment by knight666** [comment]
@Bertram25 That would be perfect. :) I will follow @hwoarangmy's advice and try to make smaller pull requests from now on. It's obvious to me now that my strategy of trying to fix as many things as possible in one go won't fly, simply because the nature of the project won't allow it. I will do my best to keep everybody informed about my progress as soon as it's available, so we can develop a vision together.

For now: this is all looking _fiiine_. I will try to get your changes integrated into my own fork as soon as possible, so I can start working on those PRs! ;)

**Comment by Bertram25** [comment]
@hwoarangmy Should be ok now :)

**Comment by Bertram25** [comment]
@knight666 Perfect. :) Eager to see what's next.

**Comment by hwoarangmy** [comment]
Tested and approved ^^
I will just rebase to be sure my PR didn't break anything here

**Comment by hwoarangmy** [comment]
Works well. I like the GUI like that ^^

**Comment by hwoarangmy** [comment]
@knight666 Basically, the rules I follow are (and most of us but since what I'm going to say has not been clearly agreed by the whole team, I will speak on my name):
A PR must compile and work. It should not introduce noticeable bugs. If someone tests and sees something, the bug must be fixed before merging. That allows to have a main branch as stable as possible (and to not break everybody else's environment)
About commits, they must have 1 and only 1 purpose
- They must not be too big. I've already made for this project commits of more than 50 files. Now that I have more experience (and that the project is more advanced), I can tell how hard it is to review such commits. If you are working on something big, try to split in smaller commits. For example, when I do something that makes me have to change the architecture, I usually do 1 first commit that changes the architecture and one other that does the wanted change.
- There should be 1 commit per purpose. We should not go for "incremental" commits and do 5 commits for 1 purpose (the 4 others being for fixing mistakes introduced by the first one). Commit history would be hard to follow otherwise.
- When a PR is released, if small mistakes are found during review, they should be fixed in a new commit. That seems to break a little bit previous rule but it allows to keep git history (the comments on the code) and helps to follow why some changes have been done.

As @Bertram25 said, this project is a spare time project and we should avoid useless work as much as possible. That's why we should go for small commits and release a PR as soon as what has been done makes the game better (even if not 100% finished). Otherwise, merge conflicts are too likely to happen.
